spec/factories/authorization_requests.rb

```diff
@@ -190,5 +191,25 @@
         )
       end
     end
+
+    trait :api_service_national do
+      type { 'AuthorizationRequest::APIServiceNational' }
+      form_uid { 'api-service-national' }
+
+      with_basic_infos
+      with_personal_data
+      with_cadre_juridique
+      with_contacts
+    end
+
+    trait :api_service_national_inscription_concours_examen do
+      type { 'AuthorizationRequest::APIServiceNational' }
+      form_uid { 'api-service-national-inscription-concours-examen' }
+    end
+
+    trait :api_service_national_obligation_service_national do
+      type { 'AuthorizationRequest::APIServiceNational' }
+      form_uid { 'api-service-national-obligation-service-national' }
+    end
```

Comment:
Je remonterai les 2 traits dans celui de `api_service_national`:

```suggestion
    trait :api_service_national do
      type { 'AuthorizationRequest::APIServiceNational' }
      form_uid { 'api-service-national' }

      with_basic_infos
      with_personal_data
      with_cadre_juridique
      with_contacts

      trait :api_service_national_inscription_concours_examen do
        form_uid { 'api-service-national-inscription-concours-examen' }
      end
  
      trait :api_service_national_obligation_service_national do
        form_uid { 'api-service-national-obligation-service-national' }
      end
    end
```

Vu que les `with_*` sont nécessaires pour que cela fonctionne. Si tu ajoutes ces 2 traits dans le test de factory tu te rendras compte que ça ne fonctionne pas.

Reply:
Ce qui veut dire que ce sera également le cas pour `api_entreprise_mdgis`

Reply:
Oui. Faudrait harden le test de factory en utilisant l'ensemble des forms pour être sûr.

Reply:
En fait, il y a une chose que je ne comprends pas, dans le cas `api_entreprise_mgdis`
```ruby
trait :api_entreprise do
      type { 'AuthorizationRequest::APIEntreprise' }
      form_uid { 'api-entreprise' }

      with_basic_infos
      with_personal_data
      with_cadre_juridique
      with_scopes
      with_contacts
    end

    trait :api_entreprise_mgdis do
      form_uid { 'api-entreprise-mgdis' }
    end
 ```
 
 Lorsque je rajoute `api_entreprise_mgdis` dans le test `spec/models/authorization_request_spec.rb`, je n'ai pas de message d'erreur 
 Pourtant si je fais un `FactoryBot.build(:authorization_request, :api_entreprise_mgdis)` en console, j'ai un message d'erreur `ActiveRecord::RecordInvalid: La validation a échoué : External est déjà utilisé(e), Organizations n'est pas valide `
 
 lorsqu'au trait de rajoute pour api_entreprise_mgdis le type, j'ai un FactoryBot fonctionnel mais le test qui failed car effectivement je dois également rajouter les with_*
 
 Je ne comprends pas trop pourquoi, on n'a pas eut d'erreur avant avec cette api ? 
 
 Et surtout, je dois rajouter à la main tous les attributs demandé dans la factorie du coup ?

Reply:
> Pourtant si je fais un FactoryBot.build(:authorization_request, :api_entreprise_mgdis) en console, j'ai un message d'erreur ActiveRecord::RecordInvalid: La validation a échoué : External est déjà utilisé(e), Organizations n'est pas valide 

Cette erreur n'a rien à voir avec les demandes mais plutôt avec des séquences mal initialisées. Faut que tu cleans ta db si tu utilises la factory en dev.

> Je ne comprends pas trop pourquoi, on n'a pas eut d'erreur avant avec cette api ?

Parce qu'on ne testait pas ce comportement (i.e. crée ce modèle avec tous les attributs)

> Et surtout, je dois rajouter à la main tous les attributs demandé dans la factorie du coup ?

Je ne comprends pas cette partie. Je vais tenter d'y répondre: oui, mais c'est déjà fait avec les `with_*` du trait `:api_entreprise`.